lib/matplotlib/cm.py

```diff
@@ -258,9 +261,14 @@ def to_rgba(self, x, alpha=None, bytes=False):
 
         # This is the normal case, mapping a scalar array:
         x = ma.asarray(x)
-        x = self.norm(x)
-        x = self.cmap(x, alpha=alpha, bytes=bytes)
-        return x
+        if norm:
+            x = self.norm(x)
+        rgba = self.cmap(x, alpha=alpha, bytes=bytes)
+        # For floating-point greyscale images, we treat negative as
+        # transparent so we copy that over to the alpha channel
+        if x.ndim == 2 and x.dtype.kind == 'f':
+            rgba[:, :, 3] = np.where(x < 0.0, 0, rgba[:, :, 3])
```

Comment:
Looks identical to  `rgba[:, :, 3][x < 0] = 0`.  No fancy indexing is required, and it is much faster than using where:
```
In [9]: xx = np.random.randn(1000, 1000, 4)

In [10]: x = xx[:, :, 0]

In [12]: timeit xx[:, :, 3][x < 0] = 0
100 loops, best of 3: 2.08 ms per loop

In [13]: timeit xx[:, :, 3] = np.where(x < 0, 0, xx[:, :, 3])
100 loops, best of 3: 9.65 ms per loop
```